tiledb/cloud/notebook.py

```diff
@@ -1,37 +1,338 @@
+        namespace,
+        storage_credential_name,
+    )
+
+
+def upload_notebook_contents(
+    ipynb_file_contents: str,
+    storage_path: str,
+    array_name: str,
+    namespace: str,
+    storage_credential_name: str,
+) -> str:
+    """
+    Uploads a notebook file to TileDB Cloud.
+    :param ipnyb_file_contents: The contents of the notebook file as a string,
+      nominally in JSON format.
+    :param storage_path: such as "s3://acmecorp-janedoe", typically from the
+      user's account settings.
+    :param array_name : name to be seen in the UI, such as "testing-upload"
+    :param namespace: such as "janedoe".
+    :param storage_credential_name: such as "janedoe-creds", typically from the
+      user's account settings.
+    :return: TileDB array name, such as "tiledb://janedoe/testing-upload".
+    """
+
+    ctx = tiledb.cloud.Ctx(
+        {"rest.creation_access_credentials_name": storage_credential_name}
+    )
+
+    tiledb_uri, array_name = _create_notebook_array(
+        storage_path,
+        array_name,
+        namespace,
+        ctx,
+    )
+
+    _write_notebook_to_array(tiledb_uri, ipynb_file_contents, ctx)
+
+    return tiledb_uri
+
+
+def _create_notebook_array(
+    storage_path: str,
+    array_name: str,
+    namespace: str,
+    ctx: tiledb.Ctx,
+    *,
+    retries: int = 0,
+) -> Tuple[str, str]:
+    """
+    Creates a new array for storing a notebook file.
+    :param storage_path: such as "s3://acmecorp-janedoe", typically from the
+      user's account settings.
+    :param array_name : name to be seen in the UI, such as "testing-upload"
+    :param namespace: such as "janedoe".
+    :param ctx: cloud context for the operation.
+    :return: tuple of tiledb_uri and array_name
+    """
+
+    if namespace in RESERVED_NAMESPACES:
+        raise ValueError(
+            f"{namespace!r} is not a valid folder to create notebooks. "
+            "Please select a proper namespace (username or organization name).",
+        )
+
+    # The array will be be 1-dimensional with domain of 0 to max uint64. We
+    # use a tile extent of 1024 bytes.
+    dom = tiledb.Domain(
+        tiledb.Dim(
+            name="position",
+            domain=(0, numpy.iinfo(numpy.uint64).max - 1025),
+            tile=1024,
+            dtype=numpy.uint64,
+            ctx=ctx,
+        ),
+        ctx=ctx,
+    )
+
+    tries = 1 + retries  # 1st + rest
+    while True:
+        try:
+            tiledb_uri, array_name = _create_notebook_array_retry_helper(
+                storage_path,
+                array_name,
+                namespace,
+                dom,
+                ctx,
+            )
+            return (tiledb_uri, array_name)
+        except tiledb.TileDBError as e:
+            if "Error while listing with prefix" in str(e):
+                # It is possible to land here if user sets wrong default S3
+                # credentials with respect to default S3 path.
+                raise ValueError(
```

Comment:
```suggestion
                raise tiledb_cloud_error.TileDBCloudError(
```

(likewise below; this is not an appropriate situation for a `ValueError`; the value is perfectly valid, but it does not work given the environment. for comparison, `os.mkdir("")` might be a `ValueError`, since making an empty directory makes no sense, but `os.mkdir("/path/to/somewhere/that/already/exists")` is *not* a ValueError.)

Reply:
> In the interest of keeping an easier-to-understand change history I would recommend the “squash and merge” option when submitting.

100% my preference as well @thetorpedodog -- thanks! :)